src/components/Form/Widgets/FileWidget.tsx

```diff
@@ -0,0 +1,216 @@
+import UploadFileIcon from '@mui/icons-material/UploadFile';
+import FilePresentIcon from '@mui/icons-material/FilePresent';
+import * as React from 'react';
+import { type DropzoneOptions, useDropzone } from 'react-dropzone';
+import {
+	Box,
+	Link,
+	Typography,
+	CircularProgress,
+	Avatar,
+	InputLabel,
+	List,
+	ListItem,
+	ListItemText,
+} from '@mui/material';
+import type { WidgetProps } from '@rjsf/utils';
+
+export interface OnFileReadSuccessParams {
+	dataUrl: string;
+	uploadedFile: File;
+}
+
+type FileWidgetProps = WidgetProps & {
+	onChange: (params?: OnFileReadSuccessParams) => void;
+	onError?: (errors: string[]) => void;
+	value?: OnFileReadSuccessParams | string;
+};
+
+export const FileWidget = ({
+	onChange,
+	onError,
+	value,
+	...props
+}: FileWidgetProps) => {
+	const [isLoading, setIsLoading] = React.useState<boolean>(false);
+	const [loadingPercentage, setLoadingPercentage] = React.useState<number>(0);
+	const [errors, setErrors] = React.useState<string[]>([]);
+
+	const accept =
+		props.uiSchema?.['ui:options']?.accept ??
+		props.uiSchema?.value?.['ui:options']?.accept;
+	const maxSize =
+		props.uiSchema?.['ui:options']?.maxSize ??
+		props.uiSchema?.value?.['ui:options']?.maxSize;
+
+	const dataUrl = (value: OnFileReadSuccessParams | string) => {
+		return typeof value === 'string' ? value : value.dataUrl;
+	};
+
+	const isImage = (value: OnFileReadSuccessParams | string) => {
+		return dataUrl(value).startsWith('data:image/');
+	};
+
+	const onDrop = React.useCallback(
+		((acceptedFiles, fileRejections) => {
+			setErrors([]);
+
+			if (fileRejections != null && fileRejections.length > 0) {
+				const errs = fileRejections[0].errors.map(
+					({ message }) => `${fileRejections[0].file.name}: ${message}`,
+				);
+				setErrors(errs);
+				onChange();
+				return onError?.(errs);
+			}
+
+			acceptedFiles.forEach((file) => {
+				const reader = new FileReader();
+				reader.onabort = () => console.log('file reading was aborted');
+				reader.onerror = () => {
+					const err = [`Failed to upload ${file.name}`];
+					onChange();
+					setErrors(err);
+					onError?.(err);
+				};
+				reader.onloadstart = () => setIsLoading(true);
+				reader.onprogress = (event) => {
+					setLoadingPercentage(Math.floor((100 * event.loaded) / event.total));
+				};
+				reader.onload = () => {
+					if (file != null && reader.result != null) {
+						setIsLoading(false);
+						onChange({
+							dataUrl: reader.result as string,
+							uploadedFile: file,
+						});
+					}
+				};
+
+				reader.readAsDataURL(file);
+			});
+		}) as NonNullable<DropzoneOptions['onDrop']>,
+		[],
```

Comment:
I think this needs to at least hold the functions we got from props

Reply:
My understanding here @myarmolinsky is that we want to use React.useCallback as documented in https://react-dropzone.js.org/ because I want to have the `isDragActive` from useDropzone().
The warning we see is a `eslint` on just because the function is not inlined and thus it can't figure it out the dependency list. However, we are not using any of the state values inside (we are just setting them) and the function does not change so I am not sure we need to pass anything on the dependency list actually

Reply:
What if the `onChange` function changes along the way? Such as if, for whatever reason, we change the `onChange` function we pass into the Form while using it? Not saying we're likely to do that, just thinking we need to be prepared for it 🤔

Reply:
done